src/cargo/ops/cargo_report/timings.rs

```diff
@@ -2,39 +2,93 @@
 
 use std::ffi::OsStr;
 use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
 use std::path::PathBuf;
 
 use annotate_snippets::Level;
 use anyhow::Context as _;
+use cargo_util::paths;
+use indexmap::IndexMap;
+use indexmap::map::Entry;
+use itertools::Itertools as _;
+use tempfile::TempDir;
 
 use crate::CargoResult;
 use crate::GlobalContext;
 use crate::core::Workspace;
+use crate::core::compiler::CompileMode;
+use crate::core::compiler::timings::CompilationSection;
+use crate::core::compiler::timings::UnitData;
+use crate::core::compiler::timings::report::RenderContext;
+use crate::core::compiler::timings::report::aggregate_sections;
+use crate::core::compiler::timings::report::compute_concurrency;
+use crate::core::compiler::timings::report::round_to_centisecond;
+use crate::core::compiler::timings::report::write_html;
 use crate::util::BuildLogger;
 use crate::util::important_paths::find_root_manifest_for_wd;
+use crate::util::log_message::LogMessage;
 use crate::util::logger::RunId;
+use crate::util::style;
 
 pub struct ReportTimingsOptions<'gctx> {
     pub log_file_path: Option<PathBuf>,
     pub id: Option<RunId>,
     pub gctx: &'gctx GlobalContext,
 }
 
+/// Collects sections data for later post-processing through [`aggregate_sections`].
+struct UnitEntry {
+    data: UnitData,
+    sections: IndexMap<String, CompilationSection>,
+    rmeta_time: Option<f64>,
+}
+
 pub fn report_timings(gctx: &GlobalContext, opts: ReportTimingsOptions<'_>) -> CargoResult<()> {
     let ws = find_root_manifest_for_wd(gctx.cwd())
         .ok()
         .and_then(|manifest_path| Workspace::new(&manifest_path, gctx).ok());
-    let Some((log, _run_id)) = select_log_file(gctx, ws.as_ref(), &opts)? else {
+    let Some((log, run_id)) = select_log_file(gctx, ws.as_ref(), &opts)? else {
         let report = [Level::INFO
             .secondary_title("no build log files found")
             .element(Level::NOTE.message("enable `-Z build-analysis` to generate logs"))];
         gctx.shell().print_report(&report, false)?;
         return Ok(());
     };
 
-    let _f = File::open(&log)
+    let ctx = prepare_context(&log, &run_id)
         .with_context(|| format!("failed to analyze log at `{}`", log.display()))?;
 
+    // If we are in a workspace,
+    // put timing reports in <target-dir>/cargo-timings/` as usual for easy access.
+    // Otherwise in a temporary directory
+    let reports_dir = if let Some(ws) = ws {
+        let target_dir = ws.target_dir();
+        let target_dir = target_dir.as_path_unlocked();
+        paths::create_dir_all_excluded_from_backups_atomic(target_dir)?;
+        let timings_dir = target_dir.join("cargo-timings");
+        paths::create_dir_all(&timings_dir)?;
+        timings_dir
+    } else {
+        TempDir::with_prefix("cargo-timings-")?.keep()
+    };
+
+    let timing_path = reports_dir.join(format!("cargo-timing-{run_id}.html"));
```

Comment:
We might also want to change the output HTML path of `--timings`, though I am not sure if it is a part of stable interface. I guess not?

* `cargo-timing-20251211T043934.741441Z.html` -> `--timing` timestamp-based file name
* `cargo-timing-2025-12-11T044138699Z-a2f1dccc7ac1667a.html` -> RunId-based file name